pkg/sql/alter_column_type.go

```diff
@@ -0,0 +1,225 @@
+	tableDesc *sqlbase.MutableTableDescriptor,
+	col *sqlbase.ColumnDescriptor,
+	t *tree.AlterTableAlterColumnType,
+) error {
+
+	usingExpressionNotSupportedErr := unimplemented.NewWithIssuef(
+		47706, "alter column type using expression is not supported")
+
+	colInIndexNotSupportedErr := unimplemented.NewWithIssuef(
+		47636, "alter column type requiring rewrite of on-disk "+
+			"data is currently not supported for columns that are part of an index")
+
+	if t.Using != nil {
+		return usingExpressionNotSupportedErr
+	}
+
+	for i := range tableDesc.Indexes {
+		for _, id := range append(
+			tableDesc.Indexes[i].ColumnIDs,
+			tableDesc.Indexes[i].ExtraColumnIDs...) {
+			if col.ID == id {
+				return colInIndexNotSupportedErr
+			}
+		}
+	}
+
+	for _, id := range append(
+		tableDesc.PrimaryIndex.ColumnIDs,
+		tableDesc.PrimaryIndex.ExtraColumnIDs...) {
+		if col.ID == id {
+			return colInIndexNotSupportedErr
+		}
+	}
+
+	currentMutationID := tableDesc.ClusterVersion.NextMutationID
+	for i := range tableDesc.Mutations {
+		mut := &tableDesc.Mutations[i]
+		if mut.MutationID < currentMutationID {
+			return unimplemented.NewWithIssuef(
+				47137, "table %s is currently undergoing a schema change", tableDesc.Name)
+		}
+	}
+
+	nameExists := func(name string) bool {
+		_, _, err := tableDesc.FindColumnByName(tree.Name(name))
+		return err == nil
+	}
+
+	shadowColName := sqlbase.GenerateUniqueConstraintName(
+		col.Name,
+		nameExists,
+	)
+
+	// Compute Expr.
+	unresolvedName := func(name string) *tree.UnresolvedName {
+		return &tree.UnresolvedName{
+			NumParts: 1,
+			Parts:    tree.NameParts{name},
+		}
+	}
+
+	newComputedExpr := tree.CastExpr{Expr: unresolvedName(col.Name), Type: t.ToType, SyntaxMode: tree.CastShort}
+	s := tree.Serialize(&newComputedExpr)
+	newColComputeExpr := &s
+
+	// May run into runtime error if DEFAULT value of column cannot be casted.
+	// In this case right now, the schema change is not rolled back.
+	hasDefault := col.HasDefault()
+	var newColDefaultExpr *string
+	if hasDefault {
+		if col.HasNullDefault() {
+			s := tree.Serialize(tree.DNull)
+			newColDefaultExpr = &s
+		} else {
+			expr, err := parser.ParseExpr(col.DefaultExprStr())
+			if err != nil {
+				return err
+			}
+			newDefaultComputedExpr := tree.CastExpr{Expr: expr, Type: t.ToType, SyntaxMode: tree.CastShort}
+			s := tree.Serialize(&newDefaultComputedExpr)
+			newColDefaultExpr = &s
+		}
+	}
+
+	id := tableDesc.NextColumnID
+	tableDesc.NextColumnID++
+
+	newCol := sqlbase.ColumnDescriptor{
+		Name:            shadowColName,
+		ID:              id,
+		Type:            *t.ToType,
+		Nullable:        col.Nullable,
+		DefaultExpr:     newColDefaultExpr,
+		UsesSequenceIds: col.UsesSequenceIds,
+		OwnsSequenceIds: col.OwnsSequenceIds,
+		ComputeExpr:     newColComputeExpr,
+	}
+
+	// Ensure shadow column is created in the same column family as the original
+	// so backfiller writes to the same column family.
+	var family sqlbase.ColumnFamilyDescriptor
```

Comment:
Can you pull this out into a utility method on a table descriptor -- i feel like this is done in alot of places.